reloader/reloader.py

```diff
@@ -65,36 +65,54 @@ def module_in_package(module):
     }
 
 
-# check the link for comments
-# https://github.com/divmain/GitSavvy/blob/599ba3cdb539875568a96a53fafb033b01708a67/common/util/reload.py
-def reload_package(pkg_name, dummy=True, verbose=True):
-    if is_dependency(pkg_name):
-        reload_dependency(pkg_name, dummy, verbose)
-        return
+def package_plugins(pkg_name):
+    return [
+        pkg_name + '.' + posixpath.basename(posixpath.splitext(path)[0])
+        for path in sublime.find_resources("*.py")
+        if posixpath.dirname(path) == 'Packages/'+pkg_name
+    ]
+
 
+def reload_package(pkg_name, dummy=True, verbose=True):
     if pkg_name not in sys.modules:
         dprint("error:", pkg_name, "is not loaded.")
         return
 
+    if is_dependency(pkg_name):
+        dependencies, packages = resolve_dependencies(pkg_name)
+    else:
+        dependencies = set()
+        packages = {pkg_name}
+
     if verbose:
         dprint("begin", fill='=')
 
-    modules = get_package_modules(pkg_name)
-
-    for m in modules:
-        if m in sys.modules:
-            sublime_plugin.unload_module(modules[m])
-            del sys.modules[m]
-
-    try:
-        with intercepting_imports(modules, verbose), \
-                importing_fromlist_aggresively(modules):
-
-            reload_plugin(pkg_name)
-    except Exception:
-        dprint("reload failed.", fill='-')
-        reload_missing(modules, verbose)
-        raise
+    # Tell Sublime to unload plugins
+    for pkg_name in packages:
+        for plugin in package_plugins(pkg_name):
+            sublime_plugin.unload_module(sys.modules[plugin])
+
+    # Unload dependencies
+    for dep_name in dependencies:
+        modules = get_package_modules(dep_name)
+        for module in modules:
+            del sys.modules[module]
+
+    # Reload packages
+    for pkg_name in packages:
+        modules = get_package_modules(pkg_name)
+        try:
+            for module in modules:
+                del sys.modules[module]
+            with intercepting_imports(modules, verbose), \
```

Comment:
I guess we should intercept both `get_package_modules(dep_name)` and `get_package_modules(pkg_name)` such that dependencies also got reloaded,

Reply:
It looks like the right thing to do is to wrap the entire plugin reload in a single `with ...` with all of the modules to reload. This results in the order of operations described in [my earlier comment](https://github.com/randy3k/AutomaticPackageReloader/issues/26#issuecomment-453271008), which I was having trouble with before.